src/Simplex/Messaging/Crypto/Ratchet.hs

```diff
@@ -0,0 +1,397 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Simplex.Messaging.Crypto.Ratchet where
+
+import Control.Monad.Except
+import Control.Monad.Trans.Except
+import Crypto.Cipher.AES (AES256)
+import qualified Crypto.Cipher.Types as AES
+import Crypto.Hash (SHA512)
+import qualified Crypto.KDF.HKDF as H
+import Data.Attoparsec.ByteString.Char8 (Parser)
+import qualified Data.Attoparsec.ByteString.Char8 as A
+import Data.ByteString.Char8 (ByteString)
+import qualified Data.ByteString.Char8 as B
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as M
+import Data.Maybe (fromMaybe)
+import Data.Word (Word16, Word32)
+import Network.Transport.Internal (decodeWord16, decodeWord32, encodeWord16, encodeWord32)
+import Simplex.Messaging.Crypto
+import Simplex.Messaging.Parsers (parseAll, parseE, parseE')
+import Simplex.Messaging.Util (tryE, (<$?>))
+
+data Ratchet a = Ratchet
+  { -- current ratchet version
+    rcVersion :: E2EEncryptionVersion,
+    -- associated data - must be the same in both parties ratchets
+    rcAD :: ByteString,
+    rcDHRs :: KeyPair a,
+    rcRK :: RatchetKey,
+    rcSnd :: Maybe (SndRatchet a),
+    rcRcv :: Maybe RcvRatchet,
+    rcMKSkipped :: Map HeaderKey SkippedMsgKeys,
+    rcNs :: Word32,
+    rcNr :: Word32,
+    rcPN :: Word32,
+    rcNHKs :: Key,
+    rcNHKr :: Key
+  }
+
+data SndRatchet a = SndRatchet
+  { rcDHRr :: PublicKey a,
+    rcCKs :: RatchetKey,
+    rcHKs :: Key
+  }
+
+data RcvRatchet = RcvRatchet
+  { rcCKr :: RatchetKey,
+    rcHKr :: Key
+  }
+
+type SkippedMsgKeys = Map Word32 MessageKey
+
+type HeaderKey = Key
+
+data MessageKey = MessageKey Key IV
+
+data ARatchet
+  = forall a.
+    (AlgorithmI a, DhAlgorithm a) =>
+    ARatchet (SAlgorithm a) (Ratchet a)
+
+-- | Input key material for double ratchet HKDF functions
+newtype RatchetKey = RatchetKey ByteString
+
+-- | Sending ratchet initialization, equivalent to RatchetInitAliceHE in double ratchet spec
+--
+-- Please note that sndPrivKey is not stored, and its public part together with random salt
+-- is sent to the recipient.
+initSndRatchet' ::
+  forall a. (AlgorithmI a, DhAlgorithm a) => PublicKey a -> PrivateKey a -> ByteString -> ByteString -> IO (Ratchet a)
+initSndRatchet' rcDHRr sPKey salt rcAD = do
+  rcDHRs@(_, pk) <- generateKeyPair' @a 0
+  let (sk, rcHKs, rcNHKr) = initKdf salt rcDHRr sPKey
```

Comment:
I think that this step should be decoupled and made outside of ratchet initialization - `initSnd/RcvRatchet` functions would be parameterized with keys that are being generated here instead. Not just because it's easier to follow double ratchet spec, but also because it allows for a different mechanism to agree on these keys. At least it could be in a separate helper function e.g. `initRatchetKeys` that would take in the DH exchange result

Reply:
maybe when it is connected, no point worrying about it in this PR